### PR TITLE
Switch to Guzzle for HTTP calls

### PR DESCRIPTION
**Pull request by elverkilde**
## Description

I have put this directly in the constructor instead of a new option. It's a fairly non-standard option IMO, so I think that makes sense.

This will be a breaking change, as support for 7.1 is removed and several deprecated functions have been removed. Will be released with new major version.

Related to #95 and #232.

## CHANGELOG

* [CHANGED] internal HTTP client to Guzzle
* [ADDED] optional client parameter to constructor
* [CHANGED] useTLS is true by default
* [REMOVED] `curl_options` from options
* [REMOVED] customer logger
* [REMOVED] host, port and timeout constructor parameters
* [REMOVED] support for PHP 7.1
* [CHANGED] lower severity level of logging to DEBUG level

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,11 +6,13 @@
     "require": {
         "php": "^7.1|^8.0",
         "ext-curl": "*",
+        "guzzlehttp/guzzle": "^7.0",
```
Comment (GrahamCampbell) [inline composer.json:9]: I'd recommend `^7.2`. In particular, the 7.0.0 release is actually broken. 7.2.0 is the first version to support PHP 8, too.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: composer.json [hunk 2]
```diff
@@ -6,11 +6,13 @@
     "require": {
         "php": "^7.1|^8.0",
```
[the remark this answers is not in the document]
Comment (elverkilde) [reply composer.json]: I notice you put 2 pipes there, is that important?

[the remark this answers is not in the document]
Comment (elverkilde) [reply composer.json]: seems like it's the same, although the double bar and spacing looks better imo. I have updated the dep. Thanks again.

File: src/Pusher.php [hunk 3]
```diff
@@ -281,20 +229,17 @@ private function validate_socket_id($socket_id)
     }
 
     /**
-     * Utility function used to create the curl object with common settings.
+     * Utility function used to generate signing headers
      *
      * @param string            $url_prefix
```
Comment (damdo) [inline src/Pusher.php:234]:
```suggestion
```

File: src/Pusher.php [hunk 4]
```diff
@@ -424,7 +300,7 @@ public static function build_auth_query_string(
 
         $auth_query_string = self::array_implode('=', '&', $params);
```
Comment (damdo) [inline src/Pusher.php:301]: We shouldn't need this now that we are returning `$params`.
If that's the case I'd also change the function name.

File: tests/unit/authQueryStringTest.php [hunk 5]
```diff
@@ -58,7 +58,13 @@ public function testGenerateSignature()
 
         $expected_to_sign = "POST\n$request_path\nauth_key=$auth_key&auth_timestamp=$time&auth_version=$auth_version&name=an_event";
         $expected_auth_signature = hash_hmac('sha256', $expected_to_sign, $auth_secret, false);
-        $expected_query_string = "auth_key=$auth_key&auth_signature=$expected_auth_signature&auth_timestamp=$time&auth_version=$auth_version&name=an_event";
+        $expected_query_string = [
```
Comment (damdo) [inline tests/unit/authQueryStringTest.php:61]: We could go with:
```suggestion
        $expected_query_params = [
```
and also change the function name.

**Review by damdo — COMMENTED** [review]
Nice one! 👍  A couple of minor comments

File: tests/unit/socketAuthTest.php [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-class PusherSocketAuthTest extends PHPUnit\Framework\TestCase
+class socketAuthTest extends PHPUnit\Framework\TestCase
```
Comment (GrahamCampbell) [inline tests/unit/socketAuthTest.php:3]: PSR-1+4 says class names must start with an uppercase letter, and the filename must match the classname.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Review by damdo — APPROVED** [review]

**Comment by caendesilva** [comment]
@elverkilde Hey! Sorry for pinging you here but I just wanted to check something. Was the removal of `curl_options` an intended change, or just a side effect?

The reason I'm asking is because it breaks self signed certificates (which are very useful for local development) when using Laravel Broadcasting.

Since `v5.0.3` no longer works on Laravel 10, this means that there is now no way to use self signed certificates when self-hosting Pusher, for example with Soketi (who documents this issue here https://docs.soketi.app/getting-started/backend-configuration/laravel-broadcasting#self-signed-certificates)

If this was an intentional removal, would you mind letting us know why it was removed? Else, would it be possible to get it added back to the package? I'm happy to try making a PR.

**Pull request merged**